pkg/evaluator/scope.go

```diff
@@ -29,3 +31,17 @@ func (s *scope) set(name string, val value) {
 	}
 	s.values[name] = val
 }
+
+func (s *scope) update(name string, val value) {
+	if s == nil {
+		panic(fmt.Errorf("%w: cannot update nil scope for %q", ErrAssignmentTarget, name))
+	}
+	if name == "_" {
+		panic(fmt.Errorf("%w: cannot update _ variable", ErrAssignmentTarget))
```

Comment:
I wonder whether it might be better to just return doing nothing for this case. Assigning to the empty var is Go is really just a no-op, so you could make that happen here. Then you would not need the checks in the various rangers.

I was going to suggest doing the same in `Set()`, but it looks like it already does just this!